hedera-node/src/main/java/com/hedera/services/context/ImmutableStateChildren.java

```diff
@@ -0,0 +1,163 @@
+import com.hedera.services.utils.EntityNum;
+import com.hedera.services.utils.EntityNumPair;
+import com.swirlds.common.AddressBook;
+import com.swirlds.merkle.map.MerkleMap;
+import com.swirlds.virtualmap.VirtualMap;
+
+import java.lang.ref.WeakReference;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A {@link StateChildren} implementation for providing immutable access to the children of a
+ * {@link ServicesState}. (Experience shows that making repeated, indirect calls to
+ * {@link com.swirlds.common.merkle.utility.AbstractNaryMerkleInternal#getChild(int)} is
+ * much more expensive, since the compiler does not seem to ever inline those calls.)
+ */
+public class ImmutableStateChildren implements StateChildren {
+	private final WeakReference<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	private final WeakReference<MerkleMap<EntityNum, MerkleTopic>> topics;
+	private final WeakReference<MerkleMap<EntityNum, MerkleToken>> tokens;
+	private final WeakReference<MerkleMap<EntityNumPair, MerkleUniqueToken>> uniqueTokens;
+	private final WeakReference<MerkleMap<EntityNum, MerkleSchedule>> schedules;
+	private final WeakReference<VirtualMap<VirtualBlobKey, VirtualBlobValue>> storage;
+	private final WeakReference<VirtualMap<ContractKey, ContractValue>> contractStorage;
+	private final WeakReference<MerkleMap<EntityNumPair, MerkleTokenRelStatus>> tokenAssociations;
+	private final WeakReference<MerkleNetworkContext> networkCtx;
+	private final WeakReference<AddressBook> addressBook;
+	private final WeakReference<MerkleSpecialFiles> specialFiles;
+	private final WeakReference<RecordsRunningHashLeaf> runningHashLeaf;
+	private final WeakReference<Map<ByteString, EntityNum>> aliases;
+	private final Instant signedAt;
+
+	private final ServicesState state;
+
+	public ImmutableStateChildren(ServicesState state) {
+		this.state = state;
+		this.signedAt = state.getTimeOfLastHandledTxn();
+		
+		accounts = new WeakReference<>(this.state.accounts());
+		topics = new WeakReference<>(this.state.topics());
+		storage = new WeakReference<>(this.state.storage());
+		contractStorage = new WeakReference<>(this.state.contractStorage());
+		tokens = new WeakReference<>(this.state.tokens());
+		tokenAssociations = new WeakReference<>(this.state.tokenAssociations());
+		schedules = new WeakReference<>(this.state.scheduleTxs());
+		networkCtx = new WeakReference<>(this.state.networkCtx());
+		addressBook = new WeakReference<>(this.state.addressBook());
+		specialFiles = new WeakReference<>(this.state.specialFiles());
+		uniqueTokens = new WeakReference<>(this.state.uniqueTokens());
+		runningHashLeaf = new WeakReference<>(this.state.runningHashLeaf());
+		aliases = new WeakReference<>(this.state.aliases());
+	}
+
+	@Override
+	public Instant signedAt() {
+		return signedAt;
+	}
+
+	@Override
+	public MerkleMap<EntityNum, MerkleAccount> accounts() {
+		final var refAccounts = accounts.get();
```

Comment:
We can use the same one-line pattern as below for the `Objects.requireNonNull()` check.